x/staking/cache/entry.go

```diff
@@ -0,0 +1,344 @@
+	storeKey := e.getStoreKey(key)
+	return store.Has(storeKey)
+}
+
+// set: caller MUST hold lock
+func (e *Entry[V, E]) setStore(store corestoretypes.KVStore, key string, bz []byte) error {
+	storeKey := e.getStoreKey(key)
+	return store.Set(storeKey, bz)
+}
+
+// deleteStore: caller MUST hold lock
+func (e *Entry[V, E]) deleteStore(store corestoretypes.KVStore, key string) error {
+	storeKey := e.getStoreKey(key)
+	return store.Delete(storeKey)
+}
+
+// count: caller MUST hold lock
+func (e *Entry[V, E]) count(store corestoretypes.KVStore) (uint64, error) {
+	prefix := e.getPrefix()
+	iter, err := store.Iterator(prefix, storetypes.PrefixEndBytes(prefix))
+	if err != nil {
+		return 0, err
+	}
+	defer iter.Close()
+
+	var count uint64
+	for ; iter.Valid(); iter.Next() {
+		count++
+	}
+	return count, nil
+}
+
+// getMetadata: caller MUST hold lock
+func (e *Entry[V, E]) getMetadata(store corestoretypes.KVStore, cdc codec.BinaryCodec) (types.CacheMetadata, error) {
+	bz, err := store.Get(e.getMetaKey())
+	if err != nil {
+		return types.CacheMetadata{}, err
+	}
+	if bz == nil {
+		return types.CacheMetadata{
+			IsDirty: true,
+			IsFull:  false,
+		}, nil
+	}
+
+	var metadata types.CacheMetadata
+	if err := cdc.Unmarshal(bz, &metadata); err != nil {
+		return types.CacheMetadata{}, err
+	}
+	return metadata, nil
+}
+
+// setMetadata: caller MUST hold lock
+func (e *Entry[V, E]) setMetadata(store corestoretypes.KVStore, cdc codec.BinaryCodec, metadata types.CacheMetadata) error {
+	bz, err := cdc.Marshal(&metadata)
+	if err != nil {
+		return err
+	}
+	return store.Set(e.getMetaKey(), bz)
+}
+
+// clear: caller MUST hold lock
+func (e *Entry[V, E]) clear(ctx context.Context, cdc codec.BinaryCodec) error {
+	store := e.storeService.OpenMemoryStore(ctx)
+	prefix := e.getPrefix()
+	iter, err := store.Iterator(prefix, storetypes.PrefixEndBytes(prefix))
+	if err != nil {
+		return err
+	}
+	defer iter.Close()
+
+	for ; iter.Valid(); iter.Next() {
+		if err := store.Delete(iter.Key()); err != nil {
+			return err
+		}
+	}
+
+	metadata, err := e.getMetadata(store, cdc)
+	if err != nil {
+		return err
+	}
+	metadata.IsFull = false
+	return e.setMetadata(store, cdc, metadata)
+}
+
+// checkReload: caller MUST hold lock
+func (e *Entry[V, E]) checkReload(ctx context.Context, cdc codec.BinaryCodec, logger func(ctx context.Context) log.Logger) error {
+	store := e.storeService.OpenMemoryStore(ctx)
+
+	metadata, err := e.getMetadata(store, cdc)
+	if err != nil {
+		return err
+	}
+
+	if metadata.IsFull {
+		return types.ErrCacheMaxSizeReached
+	}
+
+	if !metadata.IsDirty {
+		return nil
+	}
+
+	if logger != nil {
+		logger(ctx).Info(fmt.Sprintf("%s queue is dirty. Reinitializing cache from store.", e.entryType))
+	}
+
+	data, err := e.loadFromStore(ctx)
+	if err != nil {
+		return err
+	}
+
+	size := len(data)
+	if e.max > 0 && size >= int(e.max) {
+		metadata.IsFull = true
```

Comment:
Maybe change to IsDirty = true if it exceed the max size and return

and IsFull = true if it exactly reach max size